File: nonkoz/alds1_5_b.cpp
#include <iostream>
#include <cstdio>
#include <cstdlib>

int num_com = 0;
void merge(int *array, int *tmp, int left, int right, int mid){
  for(int i=left;i<= mid;i++)tmp[i] = array[i];
  for(int i=mid+1,j=right;i<=right;i++,j--)tmp[i] = array[j];  
  int i= left;
  int j= right;
  for(int k=left;k <= right;k++){
    if(tmp[i] <= tmp[j])array[k] = tmp[i++];
    else array[k] = tmp[j--];
    num_com++;
    }
}
  
void mergeSort(int *array, int *tmp,int left, int right){
  if(left >= right)return;

  int mid = (left+right)/2;
  //前半と後半の要素をそれぞれソート
  mergeSort(array, tmp, left, mid);
  mergeSort(array, tmp, mid+1,right);
  
  merge(array, tmp, left, right, mid);
  
}


int main(){
  int n;
  scanf("%d\n", &n);
  char dummy;
  int array[n];
  int tmp[n];

  for(int i=0;i<n;i++)scanf("%d%c", &array[i], &dummy);
  
  mergeSort(array, tmp, 0, n-1);
  
  for(int i=0;i<n-1;i++)printf("%d ", array[i]);
  printf("%d\n",array[n-1]);
  printf("%d\n",num_com);
  return 0;
}

File: kikunantoka/alds1_5_b.rb
INF = 2000000000
@count = 0
def merge(a, left, mid, right)
  n1 = mid - left
  n2 = right - mid
  l = []
  r = []
  (n1).times do |i|
    l[i] = a[left + i]
  end
  (n2).times do |i|
    r[i] = a[mid + i]
  end
  l[n1] = r[n2] = INF
  i = 0
  j = 0
  for k in left...right do
    @count += 1
    if l[i] <= r[j]
      a[k] = l[i]
      i = i + 1
    else
      a[k] = r[j]
      j = j + 1
    end
  end
end

def mergeSort(a, left, right)
  if left + 1 < right
    mid = (left + right) / 2
    mergeSort(a, left, mid)
    mergeSort(a, mid, right)
    merge(a, left, mid, right)
  end
end

n = gets.to_i
a = gets.chomp.split(' ').map(&:to_i)
mergeSort(a, 0, a.size())
puts a.join(' ')
puts @count
